### PR TITLE
Fix broken saturation unit tests

### PR DESCRIPTION
**Pull request by PeterCaldwell**
@pressel had hardcoded values that polysvp1 and qv_sat should match based on the Flatau et al 1992 paper these functions are based on, but there was an error in our interpretation of that paper which Kyle didn't catch. I fixed that interpretation problem a while ago, but doing so broke agreement with Kyle's hardcoded values. @ambrad commented out the resulting error so our tests would pass until I got around to fixing those hardcoded values. This PR does that.

There are a couple aspects of the current impl I'm a bit uncomfortable with, though. I'll mention them in inline comments.

### CODE REVIEW

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 1]
```diff
@@ -54,28 +54,45 @@ struct UnitWrap::UnitTest<D>::TestP3Func
 
   static void run()
```
Comment (PeterCaldwell) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp:55]: I'm creeped out that this function is just called "run". It looks like there is at least 1 other function also called "run" in this file, which seems like it should cause conflicts. Is this how we want to do things?

**Comment by PeterCaldwell** [comment]
I noticed that @pressel 's original python implementation used liquid values for ice saturation at 0 degrees C, while our C++ and F90 implementations use ice saturation at this value (which I think makes a bit more sense, though the choice is somewhat arbitrary). His choice causes an error of ~0.1 Pa to a value of 611 Pa, which is 0.01% bias, but is many orders of magnitude larger than the other polysvp1 errors we're seeing. I also notice that Kyle found the need to significantly increase error tolerance for single precision. This latter change makes sense, but I wonder whether we could/should use tighter error tolerances now that ice saturation in python and C++ are on the same page.

**Comment by PeterCaldwell** [comment]
Note: I made a mistake in a previous comment that I deleted about Kyle using the liq sat value at T=0 C while the code used the ice impl... Both F90 and C++ use the liquid value at T=0 C and I just confused myself with their logic. This is a bit of a bummer because it doesn't allow us to test the ice value directly from the paper. It also requires another tweak on my part to get tests to pass. I'm creeped out that gather-all-data passed on quartz in spite of my mistake. Tol seems set a bit coarse.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 1 above)
Comment (ambrad) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp:55]: These are static member functions of structs, so the full symbol is `static void TestName::run()`, making each unique. Same for `run_bfb`.

**Comment by ambrad** [comment]
> I'm creeped out that gather-all-data passed on quartz in spite of my mistake. Tol seems set a bit coarse.

Which tol do you mean? If this one:
```C++
Scalar tol = (util::is_single_precision<Scalar>::value || util::OnGpu<ExeSpace>::value) ? C::Tol*100 : C::Tol;
```
please do tighten it.

**Comment by ambrad** [comment]
> This is a bit of a bummer because it doesn't allow us to test the ice value directly from the paper.

Not sure what is meant here, but I'll take a guess. If you want to permit trying different values for constants, add defaulted args to the function, and make the default the physics::Constants value. Example, just making up names:
```C++
void foo(const Spack& a, Spack& b, const Real melting_temp = physics::Constants<Spack::scalar>::melting_temp) {...};
```
In a normal call, foo will use the value in the constants struct; in the test, you can provide a different value.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 1 above)
Comment (PeterCaldwell) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp:55]: Ah, thanks for pointing that out. I was led astray by the struct for saturation tests is named "TestP3Func", whose genericness led me to assume all functions in that file were in that single struct. I'll rename this struct unless anyone objects.

**Comment by PeterCaldwell** [comment]
Ok, I'm stuck. Tests are passing on Quartz but failing on Blake (both CPU machines). The code responsible for pass/fail is:
`if ( abs(sat_ice_p[s] - correct_sat_ice_p ) > tol )`. 
To debug this, I added the following lines just above this line:
`printf("sat_ice_p[s]-correct_sat_ice_p=%e\n",sat_ice_p[s] - correct_sat_ice_p);`
`printf("abs(sat_ice_p[s]-correct_sat_ice_p )=%e\n",abs(sat_ice_p[s] - correct_sat_ice_p ));`
`printf("tol = %e\n",tol);``
which returns:
> sat_ice_p[s]-correct_sat_ice_p=2.060574e-13
> tol = 1.000000e-14

so clearly the test *should* be failing. However, if I add abs to the first print statement:
`printf("abs(sat_ice_p[s]-correct_sat_ice_p)=%e\n",abs(sat_ice_p[s] - correct_sat_ice_p ));`
I get:
> abs(sat_ice_p[s]-correct_sat_ice_p )=1.000000e-14

Even weirder, if I printf the non-absolute value and then the absolute value, the absolute value print statement now returns the 2e-13 value! If I print the abs value before the non-abs value, abs returns 1e-14 and non-abs still returns 2e-13. If I create new variables for each of these quantities like:
`double ice_val = sat_ice_p[s] - correct_sat_ice_p;`
`printf("ice_val=%e\n",ice_val);`
`double abs_ice = abs(ice_val);`
`printf("abs_ice=%e\n",abs_ice);`
I get
> ice_val=2.060574e-13
>abs_ice=0.000000e+00

So it seems like maybe abs( ) is insisting on doing single-precision arithmetic? Putting double( ) around all the variables doesn't help, however, and I'm pretty sure abs( ) is appropriately overloaded to use the precision of its inputs. Or maybe this is a rounding problem (perhaps just with printed output), but 2e-13 doesn't round to 1e-14 (or vice versa). 

Does anyone have ideas what could be wrong or what I can do to get consistent answers?

**Comment by PeterCaldwell** [comment]
Ok, I got double-precision tolerance tests to pass now (yay!) but single-precision errors of 1e-4 are still cropping up (and single-precision tol = 2e-5 after I removed Kyle's hack of multiplying tol by 100). Kyle had written "The correct results were computed with double precision, so we need significantly greater tolerance for single precision." At the time, I figured that casting double-precision to double precision would essentially round at the limit of single precision so shouldn't affect agreement. I think now that Kyle was noting that polysvp1 involves ~20 calculations which could each involve different rounding at single precision and therefore causing growth of differences between results. Do people buy this? My understanding is that single precision should be accurate to 6 or 7 decimal places, so using tol=2e-5 already has substantial slop.

Also, I'm wondering whether it makes sense to set tolerances based on absolute differences (as done here) rather than relative differences. All of my failures are in vapor pressure rather than mixing ratio calculations because pressures are O(100 Pa) and mixing ratios are O(1e-3 kg/kg). Relative differences would fail if values approached zero, but that shouldn't happen in these cases. 

Thoughts? I'm particularly interested in @pressel 's reaction.

**Comment by PeterCaldwell** [comment]
Also, I'm inclined to move the saturation unit tests from "p3_unit_tests.cpp" into their own file. My understanding is that @pressel and @jgfouca started a campaign to do this a while ago. Should I do that in this PR or in a new PR?

**Comment by ambrad** [comment]
Yes, there's a condition number here that enters: error in f ~ cond(f)*eps.

**Comment by PeterCaldwell** [comment]
@ambrad - your condition number comment makes sense. My plan is to scale the 1e-14 tolerance currently being used by a factor of cond = T*f'(T)/f(T) where f is either polysvp1 or qv_sat and T is evaluated at each temperature I'm checking. Is this what you were suggesting?

**Comment by ambrad** [comment]
Yes, exactly. I think you need to do this for single precision, not double, b/c of what Kyle wrote. Or you could perturb the DP input T to be (1+eps)*T to show that the rel cond you wrote captures the error behavior in both SP and DP.

Btw, I don't like how Tol is set in the physics consts. It really should be simply
```C++
Tol = std::numeric_limits<Real>::epsilon
```
and then the end user would multiply it by something like a condition number, whether empirical or analytical. Right now the code you're working on and the conservation test are the only Tol users, so you could make that change and adjust accordingly in the PR.

**Comment by pressel** [comment]
As I recall, the so called "hack" of using a tolerance and multiplying it by a constant was done in consultation with @jgfouca and the tolerances were already set in the constants file. I agree that the condition number is the way to go, that said my experience is that using a condition number may require multiplication by an additional constant to pass the test and so it is not as cut and dry as just computing a condition number. Hopefully that constant is order 1.  I remember having a conversation with @ambrad about this at some point.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 1 above)
Comment (pressel) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp:70]: I agree TestP3Funcs should be renamed. At the time this test was added, I don't think we had a clear (at least to me) set of standards for how to name things like this, or even for that matter how tests should be distributed across files. Now that we have clear standards, we should definitely clean things like this up.

**Comment by PeterCaldwell** [comment]
I've made a bunch of changes but saturation unit tests still aren't passing. Changes:

1. made name of saturation unit test struct more descriptive
2. changed tol in physics_constants.hpp to track machine epsilon
3. analytically computed the condition # for saturation vapor pressure and mixing ratio and multiplied that by machine epsilon to get new error tolerances.
4. derived analytic formulae for saturation vapor pressure and mixing ratio (the first of which is approximate and the second is analytic) and added new tests to make sure our approx method matches expectations. 
5. quit looping over all pack entries to check agreement since packs were created by copying a scalar value a bunch of times (because functions to be checked only take packs as inputs). 
6. added debug info when tests fail.

Notes/questions:

1. conservation tests also use tol, but they continue to pass with the much tighter bounds we're not using, so I didn't add a condition number for them.
2. Approx analytic versus polysvp1 values for saturation vapor pressure are failing my (arbitrarily chosen) max tolerance of 10 Pa at 303.15 K. My approx/analytic version is 4353.995 Pa and polysvp1 yields 4245.193 Pa. These are bigger than I'd like, but it is hard to say whether it's error in polysvp1 or in my analytic approximation. The analytic method passes tests for 273 and 243 K and always passes for mixing ratio, which doesn't involve any approximations. It seems reasonable to increase my tolerance for sat vapor pressure in order to get tests passing, but bounds of 100 Pa seem pretty lame. I could also just delete the approx vapor pressure calc and keep the exact mixing ratio vals, which do pass. Opinions?
3. In addition to analytic solutions, I also recomputed values from the paper which derived the polynomial fit our model uses in python and passed them into the code (which is what Kyle had originally done). These values are within condition number x machine epsilon at double precision, but fails in ice vapor pressure for 243.15 K in single precision (model error = 1.792908e-04 and condition number x machine epsilon is 1.145070e-04). This could perhaps be explained by the fact that polysvp1 values are the result of ~20 calculations which could all cause rounding error to grow when done online (single precision) versus in python (all double precision, then rounded at the end). Should I just add a fudge factor for single precision (which is what Kyle did before)?

**Comment by ambrad** [comment]
Re: 2. Should polysvp1 be considered more accurate than your textbook derivation? If so, is there a range of T over which the textbook derivation and polysvp1 should agree well and another range where they should disagree? If so, then perhaps modify your analytical-vs-polysvp1 test to look at only the good-agreement range.

Re: 3: Right, since we're not accounting for errors in intermediates, we should expect some deviation between actual and (rel cond)*eps. The idea of using a condition number is to capture most of the error source. 1.79e-4 actual vs 1.15e-4 (rel cond)*eps is quite good agreement. A fudge factor of 2 is sufficient, it appears, and that's pretty small: rel cond and eps capture most of the behavior in an order-of-magnitude sense in this case.

**Comment by PeterCaldwell** [comment]
Great, thanks Andrew. I've implemented the fudge factor on top of the condition number. Whether the analytic or numerical implementation is more accurate is a good question, but the answer isn't particularly clear. I'll ponder a bit more.

**Comment by PeterCaldwell** [comment]
Ok, I think I've made the penultimate set of changes to this PR. Now p3_saturation_unit_tests.cpp has its own file instead of being part of p3_unit_tests.cpp. More importantly, I've cleaned up testing against an analytic solution. Ultimately I dropped saturation pressure testing because the analytic version was inexact and gave large errors that couldn't be  pinned to either the analytic or the model implementation. The mixing ratio version also gives relatively large errors (up to 5% at 303K!) relative to the exact solution. This bums me out: the vapor pressure formulation it is based on seems to have error or 0.001% based on the Flatau et al 1992 paper Fig 3 (though that paper is confusing about what it is showing or its units). I double checked the analytic implementation via direct numerical integration, so I'm pretty sure it's right. I also implemented Goff-Gratch and Murphy-Koop formulations to see if we should switch to them, but they are equally bad. All of this analysis is in scream-docs under analysis-scripts/test_qv_sat.py. Thus I'm using coarse tolerance against the exact formulation and moving on. 

One additional clean-up change which would be nice is to make separate tests for comparison against offline-calculated values using the same algorithm/approximation versus tests against the exact analytic formulation. Right now I test both things within a single test. This is sufficient but requires slightly more thought to parse errors.

**Comment by ambrad** [comment]
I lost track of saturation formula versions you're looking at, but is there a better version we could use? Sometimes approximations result from a convention that goes back decades to a time when flops weren't free. In this project, if there's a better formula that is even 2-10x more expensive than what's there, we should use it, as that cost will be in the noise.

**Comment by PeterCaldwell** [comment]
@ambrad - yeah, I've been thinking the same thing about switching from the approximate to the exact formula. I want to do that in a different PR (which I'll put off for a while because there are some other tasks which are more pressing). In addition to performance testing between our 2 methods, I will need to do some reading to make sure the exact version isn't omitting some process I can't think of which the empirical versions capture.

 the exact rather than

**Comment by PeterCaldwell** [comment]
@ambrad and @pressel - this PR has morphed significantly since you've looked at it. Would you mind a quick re-review? Thanks!

**Comment by PeterCaldwell** [comment]
Update: Further reading led me to realize that my 'exact' analytical formulation for saturation mixing ratio was actually not so exact. I've submitted #347 to fix this in comments (no changes are needed to the actual code).

**Pull request merged**